### PR TITLE
Feature enhancement

### PR DESCRIPTION
**Pull request by AndyZzzZzzZzz**
Updates:

- Enabled local execution using `MockDWaveSampler` for testing and demonstration purposes.
- Replaced `embedding_helpers` with the newly supported `minorminer` libraries for improved functionality.
- Adjusted figure formatting to align with the style used in the published paper.
- Implemented minor bug fixes and feature enhancements for better performance and usability.

### CODE REVIEW

File: tutorial_code/embed_loops.py [hunk 1]
```diff
@@ -13,57 +13,132 @@
...
+def main(max_num_emb=None):
+    from time import perf_counter
+
+    if max_num_emb is not None:
+        max_num_emb = max_num_emb
+
+    L = 2048  # L=2048 anticipate ~ 2.5 seconds on i7
+    sampler = MockDWaveSampler(topology_type="pegasus", topology_shape=[16])
+    t0 = perf_counter()
+    embeddings = embed_loops(sampler=sampler, L=L, max_num_emb=1, use_cache=False)
```
Comment (jackraymond) [inline tutorial_code/embed_loops.py:135]: Pass max_num_emb main argument to this function

File: tutorial_code/embed_loops.py [hunk 2]
```diff
@@ -13,57 +13,132 @@
...
+    t1 = perf_counter() - t0
+    if embeddings.size >= 1:
+        print(f"Loop {L} embedding successfully found in {t1} seconds")
+    else:
+        print(f"Something is wrong, {L}x{L} embedding not found")
 
 
 if __name__ == "__main__":
-    main()
+    main(max_num_emb=None)
```
Comment (jackraymond) [inline tutorial_code/embed_loops.py:144]: None is the default, just leave as () for tidiness sake.

Comment (AndyZzzZzzZzz) [reply tutorial_code/embed_loops.py:144]: I also modified for embed_lattice_square to make it same format.

File: tutorial_code/example1_2_fm_loop_correlations.py [hunk 3]
```diff
@@ -177,77 +197,113 @@ def run_experiment(param, shim, stats, embeddings, alpha_Phi=0., alpha_J=0.):
...
-    """Main function to run example
+    Args:
+        solver_name (string): option to specify sampler type. The default client QPU
+            is used by default other options are listed in Leap, to use a locally executed
+            classical placeholder for debugging select 'MockDWaveSampler'.
+        coupling (float): coupling strength on chain.
+        num_iters (int): Number of sequential programmings.
+        num_iters_unshimmed_flux (int): Number of sequential programmings without flux shimming.
+        num_iters_unshimmed_J (int): Number of sequential programmings without J shimming.
+        max_num_emb (int): Maximum number of embeddings to use per programming.
```
Comment (jackraymond) [inline tutorial_code/example1_2_fm_loop_correlations.py:261]: You forgot to add max_num_emb as an argument with default float('Inf')

File: tutorial_code/example1_2_fm_loop_correlations.py [hunk 4]
```diff
@@ -177,77 +197,113 @@ def run_experiment(param, shim, stats, embeddings, alpha_Phi=0., alpha_J=0.):
...
+        "L": 16,
+        "sampler": sampler,
+        "coupling": coupling,
+        "num_iters": num_iters,
+        "num_iters_unshimmed_flux": num_iters_unshimmed_flux,
+        "num_iters_unshimmed_J": num_iters_unshimmed_J,
     }
 
-    embeddings = embed_loops(param['L'])
+    embeddings = embed_loops(sampler=sampler, L=param["L"], use_cache=True)
```
Comment (jackraymond) [inline tutorial_code/example1_2_fm_loop_correlations.py:284]: Add max_num_emb as an argument here.

File: tutorial_code/example2_2_frustrated_loop_anneal.py [hunk 5]
```diff
@@ -203,87 +227,126 @@ def run_experiment(param, shim, stats, embeddings, alpha_Phi=0., alpha_J=0.):
...
-              all_couplings=stats['all_couplings'], frust=stats['frust'],
-              all_alpha_phi=stats['all_alpha_Phi'], all_alpha_j=stats["all_alpha_J"],
-              coupler_orbits=shim['coupler_orbits'], alpha_phi=shim['alpha_Phi'], alpha_j=shim['alpha_J'],
-              coupling=param["coupling"], L=param["L"])
-    paper_plots_example2_2(nominal_couplings=shim['nominal_couplings'],
-                           all_fbos=stats['all_fbos'],
-                           all_couplings=stats['all_couplings'],
-                           mags=stats['mags'],
-                           frust=stats['frust'])
+def main(
```
Comment (jackraymond) [inline tutorial_code/example2_2_frustrated_loop_anneal.py:272]: add 'max_num_emb' as an argument here defaulted to float('Inf')
Document it.
Pass it to embed_loops()

File: tutorial_code/example3_2_tafm_forward_anneal.py [hunk 6]
```diff
@@ -318,132 +362,204 @@ def run_experiment(param, shim, stats, embeddings, logical_bqm, alpha_Phi=0., al
...
 
-def main():
-    """Main function to run example
+def main(
+    solver_name: str = None,
+    coupling: float = 0.9,
+    num_iters: int = 800,
+    num_iters_unshimmed_flux: int = 100,
+    num_iters_unshimmed_J: int = 300,
+    max_num_emb: int = 1,
```
Comment (jackraymond) [inline tutorial_code/example3_2_tafm_forward_anneal.py:457]: This is how I would like to see max_num_emb exposed in the other example*.py files. but for embed_loops we default to float('Inf')

File: tutorial_code/example3_2_tafm_forward_anneal.py [hunk 7]
```diff
@@ -318,132 +362,204 @@ def run_experiment(param, shim, stats, embeddings, logical_bqm, alpha_Phi=0., al
...
+
+    Args:
+        solver_name (string): option to specify sampler type. The default client QPU
+            is used by default other options are listed in Leap, to use a locally executed
+            classical placeholder for debugging select 'MockDWaveSampler'.
+        coupling (float): coupling strength on chain.
+        num_iters (int): Number of sequential programmings.
+        num_iters_unshimmed_flux (int): Number of sequential programmings without flux shimming.
+        num_iters_unshimmed_J (int): Number of sequential programmings without J shimming.
+        max_num_emb (int): Maximum number of embeddings to use per programming.
```
Comment (jackraymond) [inline tutorial_code/example3_2_tafm_forward_anneal.py:474]: Can copy paste this to all example*.py main() functions

File: tutorial_code/example1_2_fm_loop_correlations.py (hunk 3 above)
Comment (AndyZzzZzzZzz) [reply tutorial_code/example1_2_fm_loop_correlations.py:261]: I thought we wanna default to 1 for this tutorial. Do we want to instead default to float('inf')?

File: tutorial_code/example3_2_tafm_forward_anneal.py (hunk 6 above)
Comment (AndyZzzZzzZzz) [reply tutorial_code/example3_2_tafm_forward_anneal.py:462]: `embed_loops` and `embed_sqaure_lattice` now has max_num_emb in main function defaulted to float('inf'). All other example scripts has max_num_emb defaulted to 1.

**Pull request closed without merging**